### DISCUSSION TITLE
TH16 Thermostat

### DISCUSSION THREAD

**Discussion by kapabel** [RavenSystem/esp-homekit-devices#1815] [MEPLHAA Scripts]
G'day.
I really tried. I really, really tried. I was diving deep into the vast ocean of possibilities of the JSON configuration. I can't figure it out, the options are too much for me to cope. I need help to make this work.

I have a solar heated jacuzzi and I want the water pump to switch ON when the water is below 38°C and OFF when it is above 42°C.

I have a TH16 ("t":22) with a DS18B20 temperature sensor ("n":3) attached to it.

Anyone could help me with the JSON config?
I really do want to understand though what each parameter means so for my learning curve and to avoid more stupid questions it would be super nice if there could be a commentary for each.

OR: Am I wrong here and even better one has to use the template from the [Device Database](https://github.com/RavenSystem/esp-homekit-devices/wiki/Devices-Database) and all the temperatures and details are setup in HomeKit?

Thanks so much for your help,
Bernhard

**Comment by xrust83** [comment]
there are two options.  1. create a thermostat with a temperature of 40 and a dead zone of 4 degrees.  but if someone changes the settings or asks Siri to set the temperature, the settings will go wrong.  2. the second option.  create a socket and a temperature sensor and rigidly through automation homeokit prescribe automation from the temperature.  which one is better for you?

**Reply by kapabel** [reply]
Right. I guess the best is option 2 and HomeKit should take care of it.

**Reply by xrust83** [reply]
{"c":{"l":13,"ct":7,"b":[{"g":0,"t":5}],"n":"Jacuzzi Heater"},"a":[{"0":{"r":[{"g":12,"v":0}]},"1":{"r":[{"g":12,"v":1}]},"t":2,"b":[{"g":0}],"s":5,"es":[{"t":22,"g":14,"n":3,"b":[],"j":5,"z":15}]}]}

**Reply by kapabel** [reply]
Thanks so much! I will try to decode this now. Many thanks!

**Reply by kapabel** [reply]
I don't understand how you managed to get this so quickly done. Is there an automated configurator somewhere? I am seriously impressed!

One more question: Why not use "ct":9 Thermostat? Isn't this exactly what this is?

**Reply by xrust83** [reply]
I shared with you the code of my water heater.  item ct is not very important, it depends on how the device will fit when added to the homekit, and then it can still be changed.

**Reply by kapabel** [reply]
I see, I will give this a go as soon as my TH16 arrives and we go from there. Thanks again for sharing your code and your time.

**Reply by xrust83** [reply]
but you have a device t:2 outlet and t:22 temperature sensor.  therefore and ct:7.  for the thermostat ct:9 device type t:21, t:25

**Reply by kapabel** [reply]
What do you think about this? It is not tested. I wrote the function as comment next to it for my understanding.
What I want to achieve is that I get in HomeKit a slider for my min and max temperature when the pump should be running (but no more than 45°C or less than 20°C). 

`
{
	"c":							// General config section
	{
		"l":13,						// Enable status LED at GPIO 13
		"ct":9,						// HomeKit Device Category: Thermostat
		"b":						// Define button to Invoke Setup Mode
			[
				{
					"g":0,			// Button GPIO 0
					"t":5			// Hold press during 8 seconds
				}
			],
		"n":"Heater"				//Custom hostname to use for the device
	},

	"a":							// Accessory section
	[
		{
			"t":21,					// Accessory type 21 (HeaterCooler)
			"g":14,					// Sensor GPIO 14
			"n":3,					// Sensor Type 3: DS18B20 Temperature sensor
			"m":20,					// Minimum Temperature 20°C
			"x":45					// Maximum Temperature 45°C
		},
		{
			"0":					// Actions triggered when HomeKit switch is set to OFF
			{
				"r":				// Digital output
				[
					{
						"g":12,		// Relay connected to GPIO 12
									//"v":0	 Value assigned to the relay OFF. Default value.
					}
				]
			},

			"1":					// Actions triggered when HomeKit switch is set to ON
			{	
				"r":				// Digital output
				[
					{
						"g":12,		// Relay connected to GPIO 12
						"v":1		// Value assigned to the relay ON
					}
				]
			},

			"b":					// Button
			[
				{
					"g":0			// Button at GPIO 0
				}
			]
		}
	]
}
`

**Reply by xrust83** [reply]
I think that you have created option 1 that I offered you, I think that your pump will not maintain a temperature not lower than 38 and not higher than 42, but will support the one you set.  for example you set 40 this thermostat will turn off i at 40.1 and turn on at 39.9 it will look like constant on / off / on / off.  you did not take into account the comfort zone or the dead zone

**Reply by xrust83** [reply]
but the most important thing when writing your json to the device, it will not work and the device will reboot in service mode.  As for the thermostat and for the socket action 1 differ.  steps 0 and 1 are required for the operation of the socket. and actions 0, 1, 3, 5 are required for the operation of the thermostat

**Reply by kapabel** [reply]
Hmm.. I would like to set the dead zone in Homekit.

Something like this:
![](https://user-images.githubusercontent.com/11726616/84404290-ecbfde80-ac06-11ea-875e-482bc5b60b10.png)

**Reply by xrust83** [reply]
this is the main mistake of those who start to create a thermostat, because in this picture you think that when the temperature drops below 20 the relay will turn on, and when it rises above 38.5 - the relay will turn off. But it doesn't work that way.
This picture means that the temperature is maintained in the range of 20-38.5, namely if the temperature drops below 20 then the heater will turn on, heat up to 20 and turn off =), and if the temperature rises (from the sun, for example) above 38.5 then turn on cooling . that is, the thermostat works if the temperature exceeds the set limits, and does nothing if the temperature is within. All thermostats in automatic mode work as for the conditioner. I could not overcome this, and could not get answers to my question on the forum

**Reply by xrust83** [reply]
{"c":{"l":13,"i":1,"n":"Heater","b":[{"g":0,"t":5}]},"a":[{"0":{"r":[{"g":12}]},"1":{"a":0},"3":{"r":[{"g":12,"v":1}]},"5":{"a":0},"t":21,"s":5,"g":14,"n":3,"j":10,"b":[{"g":0}],"d":2,"x":45,"m":15}]}

https://www.youtube.com/watch?v=d5QHQRqVPNE

**Reply by kapabel** [reply]
Fantastic! This is exactly what I need! Here is the code with some annotations for better understanding. 

```
{
	"c":							// General config section
	{
		"l":13,						// Enable status LED at GPIO 13
		"i":1,						// Invert status LED GPIO
		"n":"Heater",				// Custom hostname to use for the device
		"b":						// Define button to Invoke Setup Mode
		[
			{
				"g":0,				// Button GPIO 0
				"t":5				// Hold press during 8 seconds
			}
		]
	},

	"a":							// Accessory section
	[
		{
			"0":					// Actions triggered when HomeKit switch is set to OFF
			{
				"r":				// Digital output
				[
					{
						"g":12		// Relay connected to GPIO 12
									//"v":0	 Value assigned to the relay OFF. Default value.
					}
				]
			},
			
			"1":					// Actions triggered when HomeKit switch is set to ON
			{
				"a":0				// Copy action "0" (All OFF)
			},

			"3":					// Actions triggered when service HeaterCooler is active and heating
			{
				"r":				// Digital output
				[
					{
						"g":12,		// Relay connected to GPIO 12
						"v":1		// Value assigned to the relay ON
					}
				]
			},

			"5":					// Action to perform when there is a sensor error
			{
				"a":0				// Copy action "0" (All OFF)
			},

			"t":21,					// Accessory type 21 (HeaterCooler)
			"s":5,					// Initial start = Last state before restart
			"g":14,					// Sensor GPIO 14
			"n":3,					// Sensor Type 3: DS18B20 Temperature sensor
			"j":10,					// How often the temperature sense is read in seconds

			"b":					// Button
			[
				{
					"g":0			// Button at GPIO 0
				}
			],

			"d":2,					// Deadband/Hysteresis of service HeaterCooler
			"x":45,					// Maximum temperature that can be set
			"m":15					// Minimum temperature that can be set
		}
	]
}

```

@xrust83 : thanks a million!

**Reply by kapabel** [reply]
One more question 😬:
Will I be able to get a current temperature reading? The YouTube video is too pixelated even in HD to recognize if the current temperature is displayed. The idea is to be able to ask Siri "Hey Siri, whats the temperature in the jacuzzi?"

Is this possible in the current version or would I have to add an Extra Service in the "a" section something like:
```
,{
    "t": 22, // Service Type = Temperature Sensor
    "g": 14, // GPIO 14
    "n": 3, // Sensor Type DS18B20
    "j": 5 // Sensor polling in seconds
}
```

**Reply by xrust83** [reply]
Yes. Siri speak current temperature

**Reply by kapabel** [reply]
Fantastic! Your are a hero and thanks again for your video. Unfortunately Ukrainian isn't a language I master and Youtube doesn't automatically subtitle your video. In the meantime I managed using kapwing.com to upload your video and make some automatically transcribed Ukrainian subtitles which then in turn get automatically translated to English. Its a bad mess but I kind of understood whats going on and the video is exceptionally helpful. Thanks again and I wish you all the best regarding the current political situation.

**Reply by xrust83** [reply]
I've already subtitled some videos, but not all.  In time, I will fix it

**Comment by MukatP** [comment]
Hi, I'll join the topic. Is it possible to set the anti-freeze temperature in the Json code. For example, when the temperature drops below 5 degrees, the heating turns on and the Jacuzzi is heated to 7 degrees.
